Automattic/vip-go-mu-plugins
# Index queue system MVP

## Description

Implement MVP for asynchronous search index queuing, to provide debouncing, rate limiting, and resiliency.

This isn't complete or hooked in to automatically offload some indexing to the queue, but I think we should merge this then continue on (don't want this to get much bigger).

Known TODOs (probably future PRs)

- [ ] Cron job to process items in the queue
- [ ] Cron job to find indexable objects in the queue and create individual cron jobs to process them
- [ ] CLI commands to view/manage queue state
- [ ] Handle errors on shutdown and send only those to the queue
- [ ] Detect bulk operations and send those to the queue (maybe based on queue size)?
- [x] More tests for new queue functions (can do in this PR)

Depends on https://github.com/Automattic/ElasticPress/pull/29

## Checklist

Please make sure the items below have been covered before requesting a review:

- n/a This change has relevant documentation additions / updates (if applicable).

## Steps to Test

1. Check out PR
1. Perform some search queries
1. Ensure the new DB table (`wp_vip_search_index_queue`) is _not_ automatically created
1. Set the constant to enable async indexing: `define( 'VIP_SEARCH_ENABLE_ASYNC_INDEXING', true );`
1. Run a CLI command, like `wp vip-search health validate-counts`
1. New table should have been created automatically
1. Run the "stress test" command `wp vip-search queue stress-test`
1. In the results, it should find only 100 posts queued (despite having been re-indexed several times), and none should be "due now" (available for re-indexing now)
